File: python/3_longest_substring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        cnt = 0
        longest = 0
        memo = dict()
        i = 0
        while i < len(s):
            char = s[i]
            if char not in memo:
                memo[char] = i
                cnt += 1
            else:
                longest = max(longest, cnt)
                i = memo[char]
                memo.clear()
                cnt = 0
            i += 1
        return max(longest, cnt)

File: cpp/3_longest_substring_without_repeating_characters.cpp
#include <unordered_map>
#include <algorithm> 

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::unordered_map<char, int> memo;
        int i, cnt, longest;
        i = cnt = longest = 0;
        while(i < s.length()){
            if(memo.find(s[i]) == memo.end()){
                memo[s[i]] = i;
                cnt++;
            }
            else{
                i = memo[s[i]];
                memo.clear();
                longest = std::max(longest, cnt);
                cnt = 0;
            }
        i++;
        }
        return std::max(longest, cnt);
    }
};